### ISSUE TITLE
Rendering drawBot scripts to Coldtype viewer

### ISSUE THREAD

**Issue by stenson** [goodhertz/coldtype#7]
As suggested by @arrowtype — slightly orthogonal to how I’ve been using Coldtype (as a replacement for DrawBot rather than a modification), but the viewer app can provide a nice solution for the case of quickly previewing a drawBot script outside of the DrawBot app itself, by sending the rendered SVG to the Coldtype viewer via websockets (much like a normal coldtype render, but w/ a round-trip to an SVG `tempfile`)

Just added speculative support for this — will require re-building the viewer (via something like `yarn run start` from within the coldtype/viewer directory), and also requires a special flag passed to the coldtype cli renderer. Added a quick drawBot-syntax test file, so this command should work:

```coldtype test/test_drawbot.py -db -w```

(Not yet on pypi, so you'll need to run the command from within the coldtype repo itself for now, or install the coldtype repo into another environment locally, ala `pip install -e <path-to-coldtype-repo>`

**Comment by arrowtype** [comment]
Thanks for making this issue and adding this feature!

I had a couple of bumps in figuring this out, so I’ll note my steps here to make it simpler for others.

### Installation

I always find it slightly confusing pip installing from a repo. I did this by cloning the coldtype repo to my usual `code` folder, then I installed it into a totally separate project venv. 

For me that looks like `pip install -e /Users/stephennixon/code/coldtype`.

I think there also might be some kind of way to do this with the repo URL? Or maybe that requires an “egg” in the repo... I haven’t figured this step out fully.

### Dependencies 

After the pip install, when I try to run a DrawBot script, I run into several dependency issues, which I solved by pip installing modules.
  - `ModuleNotFoundError: No module named 'websockets'` solved with `pip install websockets`
  - `ModuleNotFoundError: No module named 'watchdog'` solved with `pip install watchdog`
  - `ModuleNotFoundError: No module named 'websocket'` solved with `pip install websocket`


### One more

```
ImportError: cannot import name 'create_connection' from 'websocket' (/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/websocket/__init__.py)
```

I found the answer [here](https://github.com/slackapi/python-slackclient/issues/57), which was `pip install websocket-client`.

### Slight design things

I had to open the ColdType app before it would work (that is, I don’t think the command started it for me, and I think it relied on me having already installed the app in the past).

I kind of expected my test script to output something that would scale down to the small size of the coldtype window, but it didn’t. So, at first, I thought it was broken, but it was actually that my test script put text at the coordinate `(500, 500)`, which was out of view. Maybe a future enhancement would be scaling the result by default, or providing an option to do so?

### Anyway

This is awesome. Thanks for making it! Hopefully my notes help someone else to use it, too.

**Comment by arrowtype** [comment]
Daaaaaang this has live reload!! I didn’t remember / appreciate that at first. Awesome.

**Comment by stenson** [comment]
Thanks for trying it out, and for the feedback! The websocket dependencies have been a stumbling block (they'd be installed if you installed coldtype with the `[renderer]` flag (i.e. `pip install -e "~/code/coldtype[renderer]"` as a drop-in replacement for the pip command you quoted above), but I think I'll add those dependencies to the core dependencies (they were only kept out because they're unnecessary in certain environments like python-in-Blender).

I've also been experimenting with a secondary mode for drawbot rendering that solves some of the issues you mentioned — currently on the `master` branch in the `test/test_drawbot_renderable.py` file — it allows you to use drawbot code but wrapped in a function, meaning the code can take advantage of the same async behaviors as the "native" Coldtype style; you'd render could render that test with `coldtype test/test_drawbot_renderable.py -w` and you'd get some of the same features as the non-drawbot rendering, like resizing the svg automatically to the bounds of the coldtype viewer app

Will update this issue as that comes together a little more & as I incorporate a few more of your suggestions!

**Comment by stenson** [comment]
Inspired by Just van Rossum’s work on drawbot-skia, I've spent the last month-ish rewriting the core GUI parts of the coldtype renderer with skia, making the library cross-platform by default and removing the need for the coldtype-viewer app entirely — all the electron code has been removed and replaced by all-python code that renders skia GL buffers to a GL window from the command-line (via glfw).

All that is mostly to say the way I was making it possible to run drawbot scripts via the coldtype CLI is no longer feasible. I think the best integration will have to come later, when drawbot-skia is further along, since the skia-ness of coldtype and drawbot-skia could be linked and coldtype would no longer need to generate images on the hard drive between renders (which is how it works now, with the cocoa version of Drawbot).

For now it is still possible to render drawbot scripts with coldtype, but they require a little bit of boilerplate, as demonstrated here: https://github.com/goodhertz/coldtype/blob/master/test/_test_drawbot_renderable.py

(Apologies to anyone following this issue, I'm mostly just writing this down for posterity)
